### ISSUE TITLE
2.3 hour duration gives 2h 17m (should be 2h 18m)

### ISSUE THREAD

**Issue by afanasy** [moment/moment#2978]
``` js
//moment.js 2.11.2
moment.duration(0.3, 'h').minutes()
//18, as expected
moment.duration(2.3, 'h').minutes()
//17
//?!?
```

I understand that it's floating point problem, but can we work around this somehow, by not converting everything to milliseconds for instance?

**Comment by icambron** [comment]
Yeah, that seems like a bug. Caused by this, I assume:

```js
> moment.duration(2.3, 'h').asMilliseconds()
8279999.999999999
```

**Comment by mj1856** [comment]
Oh JavaScript floating point numbers, you're so funny...

```javascript
2.3 * 60 * 60 * 1000 === 8280000
60 * 60 * 1000 === 3600000
2.3 * 3600000 === 8279999.999999999
```

WAT.

**Comment by mj1856** [comment]
Actually I think this one may require some more detailed thought in how we apply large constants against multiplication and division.  Probably not ideal for up-for-grabs, but if you think you have a good approach, still send a PR.  Be sure to include a unit test with at least the above example.

**Comment by afanasy** [comment]
@mj1856 Initially the approach I'd take is to store duration values with units and use them directly for conversions in .units() calls, w/o converting to ms. 

```js
//2.3 hours to minutes
//current 2.11.2 version
2.3 * 3600000 / 60000 === 137.99999999999997 //minutes - wrong
//proposal
2.3 * 60 === 138 //minutes - correct
```

but now I don't know, may be it will produce new errors with some other floating point numbers, is it really the problem with division and/or large constants?

**Comment by mj1856** [comment]
Note how there was no division in my examples.

**Comment by afanasy** [comment]
Then may be we should use some lib (or corresponding code subset) like https://github.com/infusion/Fraction.js for this?

**Comment by icambron** [comment]
Eh, floating point numbers are just painful in general. FWIW, Java behaves exactly like Matt's "WAT" example.

I propose we just change [https://github.com/moment/moment/blob/develop/src/lib/duration/constructor.js#L20](this line) to


```js
hours * 3600 * 1000
```

Does that suffice?

**Comment by mj1856** [comment]
Works for me.

**Comment by mnt** [comment]
Sorry I might be wrong but while trying:
actual result:
```
4.1 * 3600 * 1000 === 14759999.999999998
4.17123 * 1000 * 60 * 60 === 15016428.000000002
```
expected result:
```
4.1 * 3600 * 1000 === 14760000
4.17123 * 1000 * 60 * 60 === 15016428
```

**Comment by afanasy** [comment]
@mnt Looks interesting, but doesn't work for me for 
```js
(4.17123 * 100000 * 1000 * 60 * 60) / 100000
//outputs 15016428.000000002 in Google Chrome console
```
but still may be it covers more cases

**Comment by mnt** [comment]
oh im sorry you are right

**Comment by icambron** [comment]
Floating point numbers are the worst. Thinking about this a bit more, the real alternative is to just do some smarter rounding. Specifically, it's one thing to be off by 0.0000001, but it's another to round that down really far, as in the original issue, where we apparently rounded 17.999983333333333 down to 17.

**Comment by afanasy** [comment]
Yes, probably should be on this line https://github.com/moment/moment/blob/develop/src/lib/duration/get.js#L18

**Comment by NKjoep** [comment]
I don't know if it is related to `.duration` or to the DST calculation, but I have this bug:



```js

moment.duration(
  moment('2016-02-29').diff(
    moment('2016-02-24').toISOString()
  )
).days()   // returns 5 which is ok 


// then strange things happen with the DST (dst change was the 27th March)
moment.duration(
  moment('2016-03-29').diff(
    moment('2016-03-24').toISOString()
  )
).days() // return 4 which is WRONG 

```


Any help would be appreciated thanks

**Comment by icambron** [comment]
@NKjoep Please post new issues in their own tickets to keep the discussion easier to follow. I'll answer it anyway though: `diff()` returns a number of milliseconds, so it's losing the information that there's a DST there (i.e. there really *are* fewer milliseconds). What you convert that number of milliseconds to days, Moment doesn't know those millis came from two dates with a DST between them. So that's expected behavior. But you can tell Moment to take the diff in days, which *will* account for the DST:

```js
moment('2016-03-17').diff(moment('2016-03-13'), 'days'); //=> 4 (my DST is March 13)
```

**Comment by arntj** [comment]
@afanasy This gives the exact number in Chrome console.
```
parseInt(4.17123 * 100000 * 1000 * 60 * 60) / 100000

**Comment by afanasy** [comment]
@arntj yes, I think the @mnt suggestion - making milliseconds int out of float and then operating on it - is the best approach at the _moment_. While it probably will not fix all floating point issues, it will improve calculations for many similar cases. @mj1856 @icambron what do you think?

**Comment by marwahaha** [comment]
did this get finished?

**Comment by afanasy** [comment]
@marwahaha I'm not sure, because it's basically the floating point numbers standard behaviour. To fix this completely we will need to use some decimal lib to calculate things, like [big.js](https://github.com/MikeMcl/big.js/) internally.

**Comment by ashsearle** [comment]
Is there an unwritten commitment to supporting durations to sub-millisecond accuracy?

If there isn't, you could just round-off milliseconds in the `Duration` constructor:
```js
this._milliseconds = absRound(+milliseconds +
    seconds * 1e3 + // 1000
    minutes * 6e4 + // 1000 * 60
    hours * 36e5 // 1000 * 60 * 60
);
```

**Comment by afanasy** [comment]
@ashsearle that's interesting. But I'm wondering if sub milliseconds could appear unintentionally as  a by-product of some conversion 🤔

**Comment by marwahaha** [comment]
related to https://github.com/moment/moment/issues/4649

**Comment by mattjohnsonpint** [comment]
This was fixed in #2995 and released with Moment 2.13.0.  Thanks.

**Comment by afanasy** [comment]
Thank you for the update @mattjohnsonpint!
